test/native/test_runner.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "test_runner.hpp"
+#include <chrono>
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+TestRunner* TestRunner::_instance = NULL;
+
+TestRunner* TestRunner::instance()
+{
+    if (!_instance)
+    {
+        _instance = new TestRunner();
+    }
+
+    return _instance;
+}
+
+bool fileReadable(const char* file_name)
+{
+    int fd = open(file_name, O_RDONLY);
+    if (fd == -1)
+    {
+        return false;
+    }
+
+    char buf[1];
+    ssize_t r = read(fd, buf, sizeof(buf));
+    close(fd);
+    return r != -1;
+}
+
+int TestRunner::runAllTests()
+{
+    int passedTests = 0;
+    int failedTests = 0;
+    int skippedTests = 0;
+    int totalAssertions = 0;
+    int i = 1;
+    double totalDuration = 0;
+    const int totalTests = testCases().size();
+
+    const bool redirected = !isatty(fileno(stdout));
+    const char* red = redirected ? "" : "\033[31m";
+    const char* green = redirected ? "" : "\033[32m";
+    const char* yellow = redirected ? "" : "\033[33m";
+    const char* reset = redirected ? "" : "\033[0m";
+
+    bool hasOnly = false;
+    for (auto& pair : testCases())
+    {
+        hasOnly |= pair.second.only;
+    }
+    if (hasOnly)
+    {
+        printf("Only running tests marked with ONLY_TEST_CASE.\n");
+    }
+
+    for (auto& pair : testCases())
+    {
+        auto& testCase = pair.second;
+
+        printf("Running %s @ %s:%d\n", testCase.name.c_str(), testCase.filename.c_str(), testCase.lineno);
+
+        bool forceSkip = hasOnly && !testCase.only;
+        auto start = std::chrono::high_resolution_clock::now();
```

Comment:
Try not to overuse `auto` - types are *very* important for readability.
It's OK to use `auto` for lengthy types such as iterators in a narrow scope.